### ISSUE TITLE
adjoin grid cell loss compute

### ISSUE THREAD

**Issue by Nathanyaowu** [SannaPersson/YOLOv3-PyTorch#1]
![](https://user-images.githubusercontent.com/36793475/119940377-aacbac00-bfc1-11eb-92a0-bee48e6f66a3.png)
According to yolo detection thought which cell the midpoint(center_x, center_y) falls in is responsible for detect the object, but in upper code not consider the adjoin grid cell, if they also have the greater than ignore_iou_thresh, the adjoin grid cell will also compute the loss. Because the code do not set their targets[scale_idx][anchor_on_scale, i, j, 0] = -1? I am looking forward to your answer. Thank you in advance.

**Comment by SannaPersson** [comment]
I am not sure if I interpret your question correctly, but from my understanding you are wondering whether we should also consider the adjacent cells when checking for the iou_threshold. 

For a specific prediction scale and for a single target bounding box we will first check which grid cell the target has midpoint in. We will then only compare the ious in this specific grid cell. The reason to why we do not check adjacent grid cells is that they cannot contain the object's midpoint since the grid cells are non-overlapping. Each grid cell can also only predict a midpoint within the cell so an adjacent cell would not be able to predict the ground truth. In the YOLOv3 paper in the paragraph Bounding Box Prediction Redmon also writes: "If the bounding box prior is not the best but does overlap a ground truth object by
more than some threshold we ignore the prediction, following [17]." which I interpret to be similar to the implementation above. 

Since we do not consider the adjacent grid cells their targets will be set to 0 thus we will compute only a "no-object" loss for them i.e. we want them to predict that there is no object in the particular cell (since the object does not have midpoint in that cell). The -1 only indicates that this prediction should be completely ignored meaning that neither the no-object or the object loss should be computed for this particular prediction. The reason that it is important that we do not ignore the adjacent cells' predictions is that otherwise the model will not learn to accurately predict where the object's midpoint is since we will not incur a loss for where the midpoint is. 

So to summarize: We only check ious for anchors for the grid cell containing the midpoint of the object. This is since each grid cell can only predict a x- and y-position within that grid cell. Adjacent cells would therefore not be able to predict the ground truth and to me at least it does not make sense to compute a loss in that case and that is not how I interpret the paper either.

**Comment by Nathanyaowu** [comment]
Thak you for your replying, and sorry for you to got my point from my unclear presentation.But i still wondering that if an object covers 90% of the image, except the gird cell the midpint falls in,of course the adjacent grid cells will also have the anchors which they have greater than thredshold's IOU e.g. 60% or 70%, and we don't set these  targets[scale_idx][anchor_on_scale, i, j, 0] = -1,which will compute the object loss and no object loss. Becuese at adjacent grid cells they also have severals anchors with high iou with the ground truth box,but we hope this anchors to output objectness score near to 0.Does this configuraion will make the network harder to train?

**Comment by SannaPersson** [comment]
I see what you mean, and I think it is a design choice of YOLO to only have the grid cell containing the midpoint output a bounding box. In a way we are implicitly introducing the task to first classify which grid cell the object has its midpoint in. We then compute a regression loss between the actual midpoint and the predicted midpoint in that grid cell. 

It is true that other cells may also produce bounding boxes that overlap with the target with a high iou but during training we only compute an object loss for the cell containing the midpoint of the object. From my understanding this is to train the model to only output one bounding box per object. It is also part of the YOLO design to only incur a bounding box loss for the grid cells (and the specific anchors) that are assigned to a target i.e., contain the midpoint of an object. It should be noted though that even if a high iou can be achieved with predictions from other grid cells, only the cell containing the midpoint can obtain an iou of 1. This is since the predicted width and height is assumed to be symmetrical around the predicted midpoint of the object and only the grid cell containing the midpoint can predict the correct midpoint.

From my experience when training YOLOv3 it tends to sometimes make several predictions for a single object. This is likely due to it being hard to predict exactly where the midpoint is, and which anchor should be responsible for the prediction. During inference you therefore have to use non-max suppression (NMS) to remove overlapping predictions. 

I think what you are bringing up are some of the limitations with the YOLOv3 design in the way that the loss does not completely mirror the desired behavior during testing. During testing the task of predicting the grid cell of the midpoint is of less importance and we instead want to find a bounding box that maximizes the iou with the target. It seems, however, that these problems can be mitigated quite well by the use of NMS but I have also seen other implementations having success with modifying the loss.

**Comment by Nathanyaowu** [comment]
I bring up this question is because i found some other implementations of yolov3 differently. This anchors will not compute the loss  as long as the they have greater than threshold's iou with any ground truth box except the highest one, and the network can still work properly.So i think it needs detail experimentations.
And as for your implemrntation, i still confused by the Object Loss part.
object_loss = self.bce((predictions[..., 0:1][obj]), (ious * target[..., 0:1][obj]))
Why the target[..., 0:1] multiply the ious,the paper said :YOLOv3 predicts an objectness score for each bounding box using logistic regression. This should be 1 if the bounding box prior overlaps a ground truth object by more than any other bounding box prior.It means object_loss = self.bce((predictions[..., 0:1][obj]), (target[..., 0:1][obj])).

**Comment by SannaPersson** [comment]
Well yes for the anchors that is what this implementation does to. 

In all my previous answers grid cell refers to row and column in the grid that the model makes prediction in. 
For each grid cell there are several anchors connected and the loss is computed for them in the way you refer to.

As for the object loss this is my interpretation of how the object score can be computed following the equations under Direct location prediction in the YOLO9000 paper. I explain my reasoning behind this part of the loss in the paragraph "Object loss" in my article about the YOLOv3 implementation: https://sannaperzon.medium.com/yolov3-implementation-with-training-setup-from-scratch-30ecb9751cb0
